File: matt(jin)/stack/2493.cpp
#include <iostream>
#include <stack>

using namespace std;

int main() {
	int n;
	cin >> n;

	stack<int> tower;
	stack<int> tmp;
	stack<int> result;
	for (int i = 0; i < n; i++) {
		int t;
		cin >> t;
		tower.push(t);
	}
	
	for (int i = n; i > 0; i--) {
		int current = tower.top();
		tower.pop();
		int compare = 0;
		if (!tower.empty()) {
			compare = tower.top();
		}

		int howMany = 1;
		while (compare && current > compare) {
			tmp.push(compare);
			tower.pop();
			if (!tower.empty()) {
				compare = tower.top();
				howMany++;
			}
			else {
				compare = NULL;
				howMany = 0;
			}
		}

		if (compare) {
			result.push(i - howMany);
		}
		else {
			result.push(0);
		}

		while (!tmp.empty()) {
			tower.push(tmp.top());
			tmp.pop();
		}

	}

	while (!result.empty()) {
		cout << result.top() << " ";
		result.pop();
	}

	return 0;
}

File: matt(jin)/stack/2493.java
package stack;

import java.util.Scanner;

public class stack_2493 {

    public static class Tower {
        private int order;
        private int floor;

        Tower(int o, int f) {
            this.order = o;
            this.floor = f;
        }

        public int getOrder() {
            return this.order;
        }

        public int getFloor() {
            return this.floor;
        }
    }

    public static class Stack {
        private Tower[] array;
        private int size;

        Stack() {
            this.array = new Tower[10000];
            this.size = 0;
        }

        public void push(Tower tower) {
            array[size++] = tower;
        }

        public Tower top() {
            if (isEmpty()) return null;
            return array[size-1];
        }

        public Tower pop() {
            if (isEmpty()) return null;
            return array[--size];
        }

        public boolean isEmpty() {
            if(size == 0) return true;
            return false;
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        Stack stack = new Stack();
        int n = scan.nextInt();

        stack.push(new Tower(1, scan.nextInt()));
        System.out.print("0 ");

        for (int i=2; i<=n; i++) {
            int currentNum = scan.nextInt();
            Tower currentTower = new Tower(i, currentNum);

            while(!stack.isEmpty()) {
                if (currentNum < stack.top().getFloor()) {
                    System.out.print(stack.top().getOrder() + " ");
                    break;
                }
                stack.pop();
            }

            if (stack.isEmpty()) System.out.print("0 ");
            stack.push(currentTower);

        }
    }
}